modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseBatchSerializer.java

```diff
@@ -0,0 +1,485 @@
+
+        int objectIdBase = 0;
+        for (Entry<Integer, List<Lease>> entry : leasesByObjectId.entrySet()) {
+            int objectId = entry.getKey();
+            List<Lease> objectLeases = entry.getValue();
+
+            objectIdBase = writeLeasesForObject(
+                    objectId,
+                    objectLeases,
+                    minTsPhysical,
+                    commonLeasePeriod,
+                    nodesDictionary,
+                    out,
+                    objectIdBase
+            );
+        }
+    }
+
+    private static PartitionGroupId partitionedGroupIdFrom(Lease lease) {
+        return (PartitionGroupId) lease.replicationGroupId();
+    }
+
+    private static int writeLeasesForObject(
+            int objectId,
+            List<Lease> objectLeases,
+            long minTsPhysical,
+            long commonLeasePeriod,
+            NodesDictionary nodesDictionary,
+            IgniteDataOutput out,
+            int objectIdBase
+    ) throws IOException {
+        objectLeases.sort(comparing(LeaseBatchSerializer::partitionedGroupIdFrom, comparing(PartitionGroupId::partitionId)));
+
+        out.writeVarInt(objectId - objectIdBase);
+
+        int partitionCount = partitionedGroupIdFrom(objectLeases.get(objectLeases.size() - 1)).partitionId() + 1;
+        out.writeVarInt(partitionCount);
+
+        int partitionId = 0;
+        for (Lease lease : objectLeases) {
+            partitionId = writeLease(lease, partitionId, minTsPhysical, commonLeasePeriod, nodesDictionary, out);
+        }
+
+        return objectId;
+    }
+
+    private static int writeLease(
+            Lease lease,
+            int partitionId,
+            long minTsPhysical,
+            long commonLeasePeriod,
+            NodesDictionary nodesDictionary,
+            IgniteDataOutput out
+    ) throws IOException {
+        PartitionGroupId groupId = partitionedGroupIdFrom(lease);
+
+        while (partitionId < groupId.partitionId()) {
+            // It's a hole in partitionId sequence, let's write a 'dummy lease'.
+            out.write(DUMMY_LEASE_MASK);
+            partitionId++;
+        }
+
+        assert partitionId == groupId.partitionId() : "Duplicate partitionId in " + lease;
+
+        assert lease.getLeaseholder() != null && lease.getLeaseholderId() != null : lease + " doesn't have a leaseholder";
+        assert lease.getStartTime() != HybridTimestamp.MIN_VALUE : lease + " has illegal start time";
+        assert lease.getExpirationTime() != HybridTimestamp.MIN_VALUE : lease + " has illegal expiration time";
+
+        UUID leaseHolderId = lease.getLeaseholderId();
+        String proposedCandidate = lease.proposedCandidate();
+        boolean hasProposedCandidate = proposedCandidate != null;
+
+        long periodAbsolutePart = lease.getExpirationTime().getPhysical() - lease.getStartTime().getPhysical();
+        boolean hasUncommonPeriod = periodAbsolutePart != commonLeasePeriod;
```

Comment:
I doubt that many leases will have common period. Lease start time is immutable (and may be any), and expiration time is prolonged as long as the lease exists.
I would recommend to replace it with `hasUncommonExpirationTime`, and we may plan the improvement to prolong all set of leases to the same time, if at least one of them is prolonged

Reply:
Done